pkg/solana/txm/txm.go

```diff
@@ -840,25 +837,35 @@ func (txm *Txm) EstimateComputeUnitLimit(ctx context.Context, tx *solanaGo.Trans
 			}
 			err := txm.txs.OnPrebroadcastError(id, txm.cfg.TxRetentionTimeout(), txState, errType)
 			if err != nil {
-				return 0, fmt.Errorf("failed to process error %v for tx ID %s: %w", res.Err, id, err)
+				return 0, fmt.Errorf("failed to mark tx ID %s as %s: %w", id, txState, err)
 			}
 		}
 		return 0, fmt.Errorf("simulated tx returned error: %v", res.Err)
 	}
 
 	if res.UnitsConsumed == nil || *res.UnitsConsumed == 0 {
-		txm.lggr.Debug("failed to get units consumed for tx")
-		// Do not return error to allow falling back to default compute unit limit
-		return 0, nil
+		// Return error if units consumed is not estimated since it is required to verify the provided limit
+		return 0, fmt.Errorf("failed to get units consumed for tx: %s", id)
 	}
 
 	unitsConsumed := *res.UnitsConsumed
 	// Add buffer to the used compute estimate
-	computeUnitLimit := bigmath.AddPercentage(new(big.Int).SetUint64(unitsConsumed), EstimateComputeUnitLimitBuffer).Uint64()
+	estimatedComputeUnitLimit := bigmath.AddPercentage(new(big.Int).SetUint64(unitsConsumed), EstimateComputeUnitLimitBuffer).Uint64()
 	// Ensure computeUnitLimit does not exceed the max compute unit limit for a transaction after adding buffer
-	computeUnitLimit = mathutil.Min(computeUnitLimit, MaxComputeUnitLimit)
+	estimatedComputeUnitLimit = mathutil.Min(estimatedComputeUnitLimit, MaxComputeUnitLimit)
+
+	// Validate that the estimated compute unit limit does not exceed the provided limit
+	// If it does, mark the transaction as fatally errored to avoid broadcasting and wasting funds
+	if estimatedComputeUnitLimit > uint64(providedComputeUnitLimit) {
```

Comment:
Compute unit limit estimation adds a 10% buffer to the limit to account for any changes between simulation and broadcast. We're validating if the estimated limit + buffer exceeds the provided limit. Is this preferred or should this comparison happen before the buffer is added?

Reply:
If the caller has provided a limit, we should check the estimation with that limit, without adding buffer.
Buffer should only be added by us in the actual Tx, iff the estimated cu+buffer is less than the limit provided.